### PR TITLE
GH-43854: [C++] Expose the set of device types where a ChunkedArray is allocated

### PR DESCRIPTION
**Pull request by felipecrv**
### Rationale for this change

`ChunkedArray`s allow flexible allocation of arrays -- the whole array doesn't have to be allocated in huge contiguous buffers. Nothing today prevents chunked arrays from being made of chunks allocated in different devices and that is good. But we need a way to query the set of devices where a chunked array is allocated at. This PR adds that missing part.

### What changes are included in this PR?

Addition of:
- the `DeviceAllocationTypeSet` class
- `ChunkedArray::device_types()`
- `Datum::device_types()`

Moved `enum DeviceAllocationType` to the `type_fwd.h` header because `device.h` is too expensive of a header to hold this widely used `enum`.

### Are these changes tested?

Added more asserts to `chunked_array_test.cc`.

### Are there any user-facing changes?

New APIs.
* GitHub Issue: #43854

### CODE REVIEW

File: cpp/src/arrow/chunked_array.h [hunk 1]
```diff
@@ -17,6 +17,7 @@
 
 #pragma once
 
+#include <bitset>
```
Comment (felipecrv) [inline cpp/src/arrow/chunked_array.h:20]: pushing a removal of this on the next opportunity

File: cpp/src/arrow/type_fwd.h [hunk 2]
```diff
@@ -724,4 +724,23 @@ ARROW_EXPORT MemoryPool* default_memory_pool();
...
+  kROCM = 10,
+  kROCM_HOST = 11,
+  kEXT_DEV = 12,
+  kCUDA_MANAGED = 13,
+  kONEAPI = 14,
+  kWEBGPU = 15,
+  kHEXAGON = 16,
+};
+constexpr int kDeviceAllocationTypeMax = 16;
+
```
Comment (bkietz) [inline cpp/src/arrow/type_fwd.h:745]: kernel.h etc wouldn't need to change if we forward declare here
```suggestion
class DeviceAllocationTypeSet;
```

Comment (felipecrv) [reply cpp/src/arrow/type_fwd.h:745]: Not really because I'm passing `DeviceAllocationTypeSet` by value. The whole instance fits in a 64-bit word.

Comment (felipecrv) [reply cpp/src/arrow/type_fwd.h:745]: But I added a forward declaration to it nevertheless.

**Review by bkietz — APPROVED** [review]

**Pull request merged**